bothan-binance/src/worker/asset_worker.rs

```diff
@@ -0,0 +1,171 @@
+use std::sync::Weak;
+
+use rust_decimal::Decimal;
+use tokio::select;
+use tokio::sync::mpsc::Receiver;
+use tokio::time::{sleep, timeout};
+use tracing::{debug, error, info};
+
+use bothan_core::store::Store;
+use bothan_core::types::AssetInfo;
+
+use crate::api::msgs::{BinanceResponse, Data};
+use crate::api::{self, BinanceWebSocketConnection, BinanceWebSocketConnector};
+use crate::worker::error::ParseError;
+use crate::worker::types::{DEFAULT_TIMEOUT, RECONNECT_BUFFER};
+use crate::worker::BinanceWorker;
+
+pub(crate) async fn start_asset_worker(
+    worker: Weak<BinanceWorker>,
+    mut connection: BinanceWebSocketConnection,
+    mut subscribe_rx: Receiver<Vec<String>>,
+    mut unsubscribe_rx: Receiver<Vec<String>>,
+) {
+    loop {
+        select! {
+            Some(id) = subscribe_rx.recv() => handle_subscribe_recv(id, &mut connection).await,
+            Some(id) = unsubscribe_rx.recv() => handle_unsubscribe_recv(id, &mut connection).await,
+            result = timeout(DEFAULT_TIMEOUT, connection.next()) => {
+                if let Some(worker) = worker.upgrade() {
+                    match result {
+                        Err(_) => handle_reconnect(&worker.connector, &mut connection, &worker.store).await,
+                        Ok(binance_result) => handle_connection_recv(binance_result, &worker.connector, &mut connection, &worker.store).await,
+                    }
+                } else {
+                    break
```

Comment:
Should we disconnect if we know we are breaking out of this?

Reply:
added